pkg/compact/compact.go

```diff
@@ -878,26 +880,63 @@ func (cg *Group) deleteBlock(b string) error {
 
 // BucketCompactor compacts blocks in a bucket.
 type BucketCompactor struct {
-	logger     log.Logger
-	sy         *Syncer
-	comp       tsdb.Compactor
-	compactDir string
-	bkt        objstore.Bucket
+	logger      log.Logger
+	sy          *Syncer
+	comp        tsdb.Compactor
+	compactDir  string
+	bkt         objstore.Bucket
+	concurrency int
 }
 
 // NewBucketCompactor creates a new bucket compactor.
-func NewBucketCompactor(logger log.Logger, sy *Syncer, comp tsdb.Compactor, compactDir string, bkt objstore.Bucket) *BucketCompactor {
+func NewBucketCompactor(logger log.Logger, sy *Syncer, comp tsdb.Compactor, compactDir string, bkt objstore.Bucket, concurrency int) *BucketCompactor {
 	return &BucketCompactor{
-		logger:     logger,
-		sy:         sy,
-		comp:       comp,
-		compactDir: compactDir,
-		bkt:        bkt,
+		logger:      logger,
+		sy:          sy,
+		comp:        comp,
+		compactDir:  compactDir,
+		bkt:         bkt,
+		concurrency: concurrency,
 	}
 }
 
 // Compact runs compaction over bucket.
 func (c *BucketCompactor) Compact(ctx context.Context) error {
+	// Set up workers who will compact the groups when the groups are ready.
+	var (
+		errGroup, errGroupCtx = errgroup.WithContext(ctx)
+		groupChan             = make(chan *Group)
+		finishedAllGroups     bool
+		mtx                   sync.Mutex
+	)
+
+	for i := 0; i < c.concurrency; i++ {
+		errGroup.Go(func() error {
```

Comment:
eh @povilasv err groups has one big disadventage. If 2 compact will fails in the same time we will miss error from it...
see:

>  The first call to return a non-nil error cancels the group; its error will be returned by Wait.

I think that's a blocker, we would like to know all errors. Not sure, maybe just revert what it was or log error?

Reply:
Actually I think my [previous version](https://github.com/mjd95/thanos/commit/5b40b2e829029ff24e2d04f9e9e375c168e87685) had the same behaviour.  There was a buffered channel which all the goroutines could write their errors to, but as soon as the main goroutine detected an error in that channel it would return the first one it saw without reading the rest of the error channel.

In my opinion it's okay just to take the first error and return that.

Reply:
IF:

> In my opinion it's okay just to take the first error and return that.

Then it's fine :+1: